File: single-gpu/ghz_state.py
# To run with available gpus:
# ```
# python3 ghz_state.py --target nvidia
# ```
# To run with cpu (very slow):
# ```
# python3 ghz_state.py
# ```
# The CPU run is so slow that it feels
# like it's hanging.


import cudaq

def ghz_state(N):

    kernel = cudaq.make_kernel()
    q = kernel.qalloc(N)

    kernel.h(q[0])
    for i in range(N - 1):
      kernel.cx(q[i], q[i + 1])

    kernel.mz(q)
    return kernel

n = 30
print("Preparing GHZ state for", n, "qubits.")
kernel = ghz_state(n)
counts = cudaq.sample(kernel)
counts.dump()

File: single-gpu/ghz_state.cpp
// Compile and run with:
// ```
// nvq++ ghz_state.cpp -o a.out --target nvidia && ./a.out
// ```

// This example is meant to demonstrate the cuQuantum
// target and its ability to easily handle a larger number
// of qubits compared the CPU-only backend.

// Without the `--target nvidia` flag, this seems to hang, i.e.
// it takes a long time for the CPU-only backend to handle
// this number of qubits.

#include <cudaq.h>

// Define a quantum kernel with a runtime parameter
struct ghz {
  auto operator()(const int N) __qpu__ {

    // Dynamic, vector-like `qreg`
    cudaq::qreg q(N);
    h(q[0]);
    for (int i = 0; i < N - 1; i++) {
      x<cudaq::ctrl>(q[i], q[i + 1]);
    }
    mz(q);
  }
};

int main() {
  printf("Preparing GHZ state for %d qubits.\n", 30);
  auto counts = cudaq::sample(ghz{}, 30);
  counts.dump();

  return 0;
}
